<a href="https://colab.research.google.com/github/AliaBis/AliaBis/blob/main/TensorFlow.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [7]:
# экспортировать библиотеку TensorFlow
import tensorflow as tf
# библиотека для эффективной работы с массивами чисел
import numpy as np
# встроенная в Tensorflow библиотека, которая умеет обучать нейросети
from tensorflow import keras

Допустим, у нас есть два набора чисел X и Y:

X: -1 0 1 2 3 4
Y: -4 1 6 11 16 21

Мы видим, что их значения связаны по какому-то правилу. Это правило: Y = 5X + 1. Но чтобы компьютер это понял, ему нужно научиться сопоставлять входные данные — X — с результатом — Y. У него сначала могут получаться странные уравнения типа: 2X — 5, 8X + 1, 4X + 2, 5X — 1. Но, обучившись немного, он найдёт наиболее близкую к исходной формулу.

Обучается нейросеть итеративно — или поэтапно. На каждой итерации она будет предлагать алгоритм, по которому входные значения сопоставляются с результатом. Затем она проверит свои предположения, вычислив все входные данные по формуле и сравнив с настоящими результатами. Так она узнает, насколько сильно ошиблась. И уже на основе этих ошибок скорректирует формулу на следующей итерации.

Количество итераций ограничено разве что временем разработчика. Главное — чтобы нейросеть на каждом шаге улучшала свои предположения, иначе весь процесс обучения будет бессмысленным.

In [8]:
# Теперь создадим самую простую модель:
model = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])

**Sequential** — это тип нейросети, означающий, что процесс обучения будет последовательным. Это стандартный процесс обучения для простых нейросетей:в нём она сначала делает предсказания, затем тестирует их и сравнивает с результатом, а в конце — корректирует ошибки.
**keras.layers.Dense** — указывает на то, что мы хотим создать слой в нашей модели. Слой — это место, куда мы будем складывать нейроны, которые запоминают информацию об ошибках и которые отвечают за «умственные способности» нейросети. **Dense** — это тип слоя, который использует специальные алгоритмы для обучения.

В качестве аргумента нашей нейросети мы передали указания, какой именно она должна быть:

**units=1** означает, что модель состоит из одного нейрона, который будет запоминать информацию о предыдущих предположениях;
**input_shape=[1]** говорит о том, что на вход будет подаваться одно число, по которому нейросеть будет строить зависимости двух рядов чисел: X и Y.

In [9]:
# Модель мы создали, теперь давайте её скомпилируем:
model.compile(optimizer='sgd', loss='mean_squared_error')

Здесь появляются два важных для машинного обучения элемента: функция оптимизации и функция потерь. Обе они нужны, чтобы постепенно стремиться к более точным результатам.

Функция потерь анализирует, насколько правильно нейросеть дала предсказание. А функция оптимизации исправляет эти предсказания в сторону более корректных результатов.

Мы использовали стандартные функции для большинства моделей — sgd и mean_squared_error. sgd — это метод оптимизации, который работает на формулах математического анализа. Он помогает скорректировать формулу, чтобы прийти к правильной. mean_squared_error — это функция, которая вычисляет, насколько сильно отличаются полученные результаты по формуле, предложенной нейросетью, от настоящих результатов. Эта функция тоже участвует в корректировке формулы.

In [10]:
# Теперь давайте зададим наборы данных:
xs = np.array([-1.0, 0.0, 1.0, 2.0, 3.0, 4.0], dtype=float)
ys = np.array([-4.0, 1.0, 6.0, 11.0, 16.0, 21.0], dtype=float)
# Как видно, это обычные массивы чисел, которые мы передадим модели на обучение:

In [11]:
model.fit(xs, ys, epochs=500)

Epoch 1/500
1/1 [==============================] - 0s 487ms/step - loss: 140.1312
Epoch 2/500
1/1 [==============================] - 0s 16ms/step - loss: 110.2791
Epoch 3/500
1/1 [==============================] - 0s 11ms/step - loss: 86.7922
Epoch 4/500
1/1 [==============================] - 0s 10ms/step - loss: 68.3133
Epoch 5/500
1/1 [==============================] - 0s 7ms/step - loss: 53.7744
Epoch 6/500
1/1 [==============================] - 0s 9ms/step - loss: 42.3352
Epoch 7/500
1/1 [==============================] - 0s 9ms/step - loss: 33.3349
Epoch 8/500
1/1 [==============================] - 0s 9ms/step - loss: 26.2533
Epoch 9/500
1/1 [==============================] - 0s 7ms/step - loss: 20.6813
Epoch 10/500
1/1 [==============================] - 0s 8ms/step - loss: 16.2970
Epoch 11/500
1/1 [==============================] - 0s 5ms/step - loss: 12.8470
Epoch 12/500
1/1 [==============================] - 0s 16ms/step - loss: 10.1323
Epoch 13/500
1/1 [=======================

Функция **fit** как раз занимается обучением. Она берёт набор входных данных — xs — и сопоставляет с набором правильных результатов — ys. И так нейросеть обучается в течение 500 итераций — epochs=500. Мы использовали 500 итераций, чтобы наверняка прийти к правильному результату. Суть простая: чем больше итераций обучения, тем точнее будут результаты (однако улучшение точности с каждым повтором будет всё меньше и меньше).

##На каждой итерации модель проходит следующие шаги:

- берёт весь наш набор входных данных;
- пытается сделать предсказание для каждого элемента;
- сравнивает результат с корректным результатом;
- оптимизирует модель, чтобы давать более точные прогнозы.

Можно заметить, что на каждой итерации TensorFlow выводит, насколько нейросеть сильно ошиблась — loss. Если это число уменьшается, то есть стремится к нулю, значит, она действительно обучается и с каждым шагом улучшает свои прогнозы.

Теперь давайте что-нибудь предскажем и поймём, насколько точно наша нейросеть обучилась:

In [13]:
print(model.predict([10.0]))
# ожидаем,что будет число 51. Это 5х-1

1/1 [==============================] - 0s 39ms/step
[[50.993446]]


Мы вызываем у модели метод predict, который получает на вход элемент для предсказания. Результат будет таким: 
1/1 [==============================] - 0s 39ms/step
[[50.993446]]
Модель сделала предсказание, но оно оказалось не совсем точным

Получилось странно — мы ожидали, что будет число 51 (потому что подставили 10 в выражение 5X + 1) — но на выходе нейросеть выдала число 50.98739. А всё потому, что модель нашла очень близкую, но не до конца точную формулу — например, 4.891X + 0.993. Это одна из особенностей машинного обучения.

##Что запомнить
##Библиотека TensorFlow — это инструмент для создания и обучения нейросетей. При этом вам не нужно углублённо знать высшую математику, чтобы писать простые модели.
##Машинное обучение — это когда мы даём компьютеру входные данные и результаты и просим его понять между ними зависимость. Нам неважно, как он до этого додумается. Главное — точность.
##Обучение нейросети проходит в три этапа: подготовка данных, создание и компиляция модели, само обучение.